test/TokenPaymaster.test.js

```diff
@@ -0,0 +1,228 @@
+
+    recipient = await TestProxy.new()
+    await recipient.setRelayHub(hub.address)
+
+    // approve uniswap to take our tokens.
+    await token.approve(uniswap.address, -1)
+
+    sharedRelayRequestData = {
+      senderAddress: from,
+      encodedFunction: recipient.contract.methods.test().encodeABI(),
+      senderNonce: '1',
+      target: recipient.address,
+      pctRelayFee: '1',
+      baseRelayFee: '0',
+      gasPrice: await web3.eth.getGasPrice(),
+      gasLimit: 1e6.toString(),
+      relayAddress: from,
+      paymaster: paymaster.address
+    }
+  })
+
+  context('#TestUniswap', () => {
+    it('check exchange rate', async () => {
+      assert.equal((await uniswap.getTokenToEthOutputPrice(2e10)).toString(), 4e10)
+      assert.equal((await uniswap.getTokenToEthInputPrice(2e10)).toString(), 1e10)
+    })
+
+    it.skip('swap token to eth', async () => {
+      await token.mint(10e18.toString())
+      const ethBefore = await web3.eth.getBalance(from)
+      const tokensBefore = await token.balanceOf(from)
+      // zero price for easier calculation
+      await uniswap.tokenToEthSwapOutput(2e18.toString(), -1, -1, { gasPrice: 0 })
+      const ethAfter = await web3.eth.getBalance(from)
+      const tokensAfter = await token.balanceOf(from)
+
+      assert.equal((tokensAfter - tokensBefore) / 1e18, -4)
+      assert.equal((ethAfter - ethBefore) / 1e18, 2)
+    })
+
+    it('swap and transfer', async () => {
+      await token.mint(10e18.toString())
+      const target = '0x' + '1'.repeat(40)
+      const tokensBefore = await token.balanceOf(from)
+      const ethBefore = await web3.eth.getBalance(target)
+      await uniswap.tokenToEthTransferOutput(2e18.toString(), -1, -1, target)
+      const tokensAfter = await token.balanceOf(from)
+
+      const ethAfter = await web3.eth.getBalance(target)
+      assert.equal((tokensAfter - tokensBefore) / 1e18, -4)
+      assert.equal((ethAfter - ethBefore) / 1e18, 2)
+    })
+  })
+
+  context('#acceptRelayedCall', async () => {
+    it('fail if not enough balance', async () => {
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData
+      })
+      assert.deepEqual(await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6)), {
+        code: 99,
+        msg: 'balance too low'
+      })
+    })
+
+    it('fund recipient', async () => {
+      await token.mint(5e18.toString())
+      await token.transfer(recipient.address, 5e18.toString())
+    })
+
+    it('fail if no approval', async () => {
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData
+      })
+      assert.deepEqual(await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6)), {
+        code: 99,
+        msg: 'allowance too low'
+      })
+    })
+
+    it('recipient.approve', async () => {
+      await recipient.execute(token.address, token.contract.methods.approve(paymaster.address, -1).encodeABI())
+    })
+
+    it('succeed?', async () => {
```

Comment:
'it' description should make it clear what the single failed test was doing. If this test fails, you'll see a big red 
### succeed?

Reply:
ok. now it "should.."